### PR TITLE
napi test driver setup

### PR DESCRIPTION
**Pull request by FloVanGH**
* Added napi test driver
* Added missing pieces to get tests green
* Fixed stuff in napi to get tests green
* Added additional napi tests

### CODE REVIEW

File: tests/cases/types/brush.slint [hunk 1]
```diff
@@ -58,12 +58,12 @@ assert_eq!(t.get_color_brush().color(), green);
 
 ```js
 var t = new slint.Test({});
-assert.equal(t.color_brush, "#0000ffff");
+assert.equal(t.color_brush.toString(), "#0000ffff");
```
Comment (ogoffart) [inline tests/cases/types/brush.slint:61]: Is that needed or is the conversion not implicit?

File: api/napi/index.ts [hunk 2]
```diff
@@ -152,6 +152,8 @@ export interface ComponentHandle {
     show();
     hide();
     get window(): Window;
+    send_mouse_click(x: number, y: number);
+    send_keyboard_string_sequence(s: String);
```
Comment (ogoffart) [inline api/napi/index.ts:156]: Is there a way to only have that for the tests?  I actually don't think it should be part of the public interface

File: tests/cases/types/brush.slint (hunk 1 above)
Comment (FloVanGH) [reply tests/cases/types/brush.slint:61]: It does it implicit on `assert` but not with `assert.equal`.

File: api/napi/src/interpreter/value.rs [hunk 3]
```diff
@@ -105,12 +108,39 @@ pub fn to_value(env: &Env, unknown: JsUnknown, typ: Type) -> Result<Value> {
             Ok(Value::Bool(js_bool.get_value()?))
         }
         Type::Color => {
-            let js_color: JsExternal = unknown.coerce_to_object()?.get("color")?.unwrap();
-            Ok(Value::Brush(Brush::from(env.get_value_external::<Color>(&js_color)?.clone())))
+            if unknown.get_type()?.eq(&ValueType::String) {
```
Comment (tronical) [inline api/napi/src/interpreter/value.rs:111]: Are you sure about the `?` here? It seems to me that if that fails, we're not even going to try the "native" color code path in the else. Shouldn't we?

I think a more robust overall logic would be this:

1. Try to coerce the unknown to an object, extract the color property, attempt to extract the `Color` external.
2. If any of the above steps fail, attempt to coerce to a string and then try to the CSS parsing.

The same applies to the brush code. Perhaps there's a way to avoid the code duplication by sharing the code in a helper function?

File: api/napi/src/types/brush.rs [hunk 4]
```diff
@@ -151,4 +161,9 @@ impl JsBrush {
     pub fn brush(&self) -> External<Brush> {
         External::new(self.inner.clone())
     }
+
+    #[napi]
+    pub fn to_string(&self) -> String {
+        self.color().to_string()
```
Comment (tronical) [inline api/napi/src/types/brush.rs:167]: I suggest to add a comment here that this is wrong for brushes that aren't solid colors. Evidently a gradient brush should not convert to a string like this. Alternatively, perhaps an exception should be thrown, that this is not implemented yet?

File: examples/todo/ui/todo.slint [hunk 5]
```diff
@@ -114,6 +114,7 @@ component MainWindow inherits Window {
         HorizontalBox {
             padding: 0px;
             alignment: end;
+
```
Comment (tronical) [inline examples/todo/ui/todo.slint:117]: Unrelated change?

File: tests/driver/napi/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+publish = false
+
+[[bin]]
+path = "main.rs"
+name = "test-driver-napi"
+
+[dev-dependencies]
+test_driver_lib = { path = "../driverlib" }
+# Require `artifact`  dependencies tracked by rust RFC 3028
+#slint-node = { path = "../../../api/napi", artifact = ["cdylib"] }
```
Comment (tronical) [inline tests/driver/napi/Cargo.toml:22]: It looks like that despite this being in a comment for future adaptation, the path and name should be changed.

File: tests/driver/napi/build.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    target_dir.pop();
+    target_dir.pop();
+    target_dir.pop();
+
+    let nodejs_native_lib_name = {
+        let (prefix, suffix) = os_dylib_prefix_and_suffix();
+        format!("{}slint_node_native.{}", prefix, suffix)
+    };
+    println!(
+        "cargo:rustc-env=SLINT_NODE_NATIVE_LIB={}",
```
Comment (tronical) [inline tests/driver/napi/build.rs:29]: Why is this nodejs specific logic included here? The napi generated loader takes care of loading the correct platform-specific file, doesn't it?

File: tests/driver/napi/nodejs.rs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+lazy_static::lazy_static! {
+    static ref NODE_API_JS_PATH: PathBuf = {
+        let  node_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../../api/napi");
+
+        // On Windows npm is 'npm.cmd', which Rust's process::Command doesn't look for as extension, because
+         // it tries to emulate CreateProcess.
+         let npm = which::which("npm").unwrap();
+
+        // Ensure TypeScript is installed
```
Comment (tronical) [inline tests/driver/napi/nodejs.rs:15]: This comment seems wrong, as npm install for napi's package.json does more.

File: tests/driver/napi/nodejs.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+       std::process::Command::new(npm.clone())
+            .arg("install")
+            .arg("--no-audit")
+            .current_dir(node_dir.clone())
+            .stdout(std::process::Stdio::piped())
+            .stderr(std::process::Stdio::piped())
+            .output()
+            .map_err(|err| format!("Could not launch npm install: {}", err)).unwrap();
+
+        // Build the .js file of the NodeJS API from the .ts file
```
Comment (tronical) [inline tests/driver/napi/nodejs.rs:25]: Why is this step needed for the napi port?

File: examples/todo/ui/todo.slint (hunk 5 above)
Comment (FloVanGH) [reply examples/todo/ui/todo.slint:117]: yes only formatting.

File: tests/driver/napi/nodejs.rs (hunk 9 above)
Comment (tronical) [reply tests/driver/napi/nodejs.rs:25]: I'm asking because the previous step runs "npm install", which in `package.json` runs `npm build`. So it seems unnecessary to run `npm build` here again.

Why did the tests not work without it?

Comment (FloVanGH) [reply tests/driver/napi/nodejs.rs:25]: Ok after a clean it works now.

File: tests/driver/napi/nodejs.rs (hunk 8 above)
Comment (FloVanGH) [reply tests/driver/napi/nodejs.rs]: I have updated the comment.

File: tests/driver/napi/build.rs (hunk 7 above)
Comment (FloVanGH) [reply tests/driver/napi/build.rs]: Removed.

File: tests/driver/napi/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+publish = false
+
+[[bin]]
+path = "main.rs"
+name = "test-driver-napi"
+
+[dev-dependencies]
+test_driver_lib = { path = "../driverlib" }
+# Require `artifact`  dependencies tracked by rust RFC 3028
+#slint-node = { path = "../../../api/node", artifact = ["cdylib"] }
```
Comment (tronical) [inline tests/driver/napi/Cargo.toml:22]:
```suggestion
#slint-napi = { path = "../../../api/napi", artifact = ["cdylib"] }
```

File: tests/driver/napi/build.rs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+fn os_dylib_prefix_and_suffix() -> (&'static str, &'static str) {
+    if cfg!(target_os = "windows") {
+        ("", "dll")
+    } else if cfg!(target_os = "macos") || cfg!(target_os = "ios") {
+        ("lib", "dylib")
+    } else {
+        ("lib", "so")
+    }
+}
+
```
Comment (tronical) [inline tests/driver/napi/build.rs:16]:
```suggestion
```

File: api/napi/index.ts (hunk 2 above)
Comment (FloVanGH) [reply api/napi/index.ts]: done

File: api/napi/src/interpreter/value.rs (hunk 3 above)
Comment (FloVanGH) [reply api/napi/src/interpreter/value.rs]: done

File: api/napi/Cargo.toml [hunk 12]
```diff
@@ -25,6 +25,7 @@ i-slint-compiler = { workspace = true, features = ["default"] }
 i-slint-core = { workspace = true, features = ["default"] }
 slint-interpreter = { workspace = true, features = ["default", "display-diagnostics"] }
 spin_on = "0.1"
+css-color-parser2 = "1.0.1"
```
[the remark this answers is not in the document]
Comment (FloVanGH) [reply api/napi/Cargo.toml]: done

File: api/napi/src/interpreter/value.rs [hunk 13]
```diff
@@ -105,12 +108,52 @@ pub fn to_value(env: &Env, unknown: JsUnknown, typ: Type) -> Result<Value> {
...
+                return Ok(Value::Brush(js_color.clone()));
+            }
+
+            if let Some(js_brush) = env
+                .get_reference_value::<JsObject>(&color_ref)
+                .ok()
+                .and_then(|js_object| js_object.coerce_to_string().ok())
+                .and_then(|string| string_to_brush(string).ok())
+            {
+                return Ok(js_brush);
```
Comment (tronical) [inline api/napi/src/interpreter/value.rs:127]: Not strictly related to this patch, as this happened before as well, but when `to_value` is called with `Type::Color`, the function does not return a `Value::Color` but a `Value::Brush`. That seems wrong, and the nodejs bindings seems to have gotten it "right" by doing the opposite: Always returning a `Value::Color` even when a `Brush` is requested - although that conversion is safe to do.

**Review by tronical — APPROVED** [review]
